### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by andriy-shymkiv**
[DEMO LINK](https://andriy-shymkiv.github.io/react_todo-app-with-api/)

### CODE REVIEW

**Review by referee1988 — CHANGES_REQUESTED** [review]
**1. When there is no TODO, there should be no switch.**
image

**2. Loader does not appear when adding TODO**
https://user-images.githubusercontent.com/105992827/208402075-f15f569b-b18c-4708-b287-ec7d7ee43317.mp4

**3. When deleting several TODOs, the Loader does not load for all TODOs**
https://user-images.githubusercontent.com/105992827/208403190-bc3798e1-89ef-49a1-ad96-073e8aa7251c.mp4

**4. The status switcher should change the statuses of all Active TODOs to Completed**
https://user-images.githubusercontent.com/105992827/208403990-e574a521-a33c-4b5f-985e-bea27d958304.mp4

**5.  When you turn on the edit mode, the field for changing the name of the TODO should be focused.
Also, the editing mode should be canceled when editing the next TODO is started.**
https://user-images.githubusercontent.com/105992827/208404611-6e67c808-4361-4b07-ba29-54d7b45bcc3e.mp4

**6. There are constant requests to the server for a complete TODO list**
https://user-images.githubusercontent.com/105992827/208405511-dc69b572-565b-457b-b60d-382144e54734.mp4

**Review by referee1988 — CHANGES_REQUESTED** [review]
1. In TODO renaming mode, when you press Escape, editing should end, but nothing happens to you.

2. In TODO renaming mode, when you click the mouse outside the input field, renaming should occur (as with Enter), and your changes are canceled.
 https://user-images.githubusercontent.com/105992827/208968556-72a64536-1b62-4ab1-9ae4-b1e1d82e65b6.mp4

3. Loader does not work correctly when renaming several TODOs in succession
https://user-images.githubusercontent.com/105992827/208970884-d0279d59-a89a-4b98-afcb-788702f4bef1.mp4

4. When changing TODO, it is not necessary to download all TODO from the server again, it is enough to change the corresponding TODO in useState.
https://user-images.githubusercontent.com/105992827/208970273-5f48ed7b-c1cd-49ca-b42d-a3f726a8e93a.mp4

File: src/App.tsx [hunk 1]
```diff
@@ -1,223 +1,250 @@
...
+
+        if (todosFromServer && todosFromServer.length > 0) {
+          setShowFooter(true);
+        }
+
+        if (todosFromServer && todosFromServer.every(todo => todo.completed)) {
+          setToggleButton(true);
+        }
+
+        if (todosFromServer && todosFromServer.some(todo => !todo.completed)) {
```
Comment (illiarerun) [inline src/App.tsx:62]: that's basically the same check as previous, you can just use `else`

File: src/App.tsx [hunk 2]
```diff
@@ -1,223 +1,250 @@
...
+
   const user = useContext(AuthContext);
-  const newTodoField = useRef<HTMLInputElement>(null);
+  const {
+    None,
+    NoTodos,
+    Update,
+    Delete,
+    Add,
+  } = ErrorType;
```
Comment (illiarerun) [inline src/App.tsx:43]: you're creating new variables on each rerender, move it out of the component

File: src/components/Footer/Footer.tsx [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        </a>
+      </nav>
+
+      <button
+        data-cy="ClearCompletedButton"
+        type="button"
+        className="todoapp__clear-completed"
+        onClick={() => handleClearCompleted()}
+      >
+        {hasCompletedTodos ? 'Clear complited' : ''}
```
Comment (illiarerun) [inline src/components/Footer/Footer.tsx:159]:
```suggestion
        {hasCompletedTodos ? 'Clear completed : ''}
```

**Review by illiarerun — APPROVED** [review]

**Pull request closed without merging**